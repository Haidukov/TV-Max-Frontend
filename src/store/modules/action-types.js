export * from './shows/action-types';
export * from './episodes/action-types';
