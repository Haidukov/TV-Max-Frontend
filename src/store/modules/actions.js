export * from './shows/actions';
export * from './episodes/actions';
