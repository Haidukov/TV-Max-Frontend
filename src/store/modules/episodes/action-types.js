export const FETCH_EPISODES_REQUEST = 'FETCH_EPISODES_REQUEST';
export const FETCH_EPISODES_SUCCESS = 'FETCH_EPISODES_SUCCESS';
export const FETCH_EPISODES_ERROR = 'FETCH_EPISODES_ERROR';

export const FETCH_EPISODE_REQUEST = 'FETCH_EPISODE_REQUEST';
export const FETCH_EPISODE_SUCCESS = 'FETCH_EPISODE_SUCCESS';
export const FETCH_EPISODE_ERROR = 'FETCH_EPISODE_ERROR';
