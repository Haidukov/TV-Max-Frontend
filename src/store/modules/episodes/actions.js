import {
  FETCH_EPISODES_ERROR,
  FETCH_EPISODES_REQUEST,
  FETCH_EPISODES_SUCCESS,
  FETCH_EPISODE_ERROR,
  FETCH_EPISODE_REQUEST,
  FETCH_EPISODE_SUCCESS
} from '../action-types';

export const fetchEpisodesRequest = id => ({
  type: FETCH_EPISODES_REQUEST,
  payload: { id }
});

export const fetchEpisodesSuccess = (id, episodes) => ({
  type: FETCH_EPISODES_SUCCESS,
  payload: {
    id,
    episodes
  }
});

export const fetchEpisodesError = (id, error) => ({
  type: FETCH_EPISODES_ERROR,
  payload: {
    id,
    error
  }
});

export const fetchEpisodeRequest = id => ({
  type: FETCH_EPISODE_REQUEST,
  payload: { id }
});

export const fetchEpisodeSuccess = (id, data) => ({
  type: FETCH_EPISODE_SUCCESS,
  payload: {
    id,
    data
  }
});

export const fetchEpisodeError = (id, error) => ({
  type: FETCH_EPISODE_ERROR,
  payload: {
    id,
    error
  }
});
