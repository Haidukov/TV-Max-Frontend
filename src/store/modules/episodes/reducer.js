import produce from 'immer';
import {
  FETCH_EPISODES_ERROR,
  FETCH_EPISODES_REQUEST,
  FETCH_EPISODES_SUCCESS,
  FETCH_EPISODE_ERROR,
  FETCH_EPISODE_REQUEST,
  FETCH_EPISODE_SUCCESS
} from '../action-types';

const initialState = {
  byId: {},
  byShow: {}
};

const reducer = produce((draft, action) => {
  switch (action.type) {
    case FETCH_EPISODES_REQUEST: {
      if (!draft.byShow[action.payload.id]) {
        draft.byShow[action.payload.id] = {
          ids: []
        };
      }
      draft.byShow[action.payload.id].isLoading = true;
      draft.byShow[action.payload.id].error = null;
      break;
    }
    case FETCH_EPISODES_SUCCESS: {
      if (!draft.byShow[action.payload.id]) {
        draft.byShow[action.payload.id] = {
          ids: []
        };
      }
      draft.byShow[action.payload.id].isLoading = false;
      draft.byShow[action.payload.id].ids = [];
      action.payload.episodes.forEach(episode => {
        if (!draft.byId[episode.id]) {
          draft.byId[episode.id] = {};
        }
        draft.byId[episode.id].data = episode;
        draft.byShow[action.payload.id].ids.push(episode.id);
      });
      break;
    }
    case FETCH_EPISODES_ERROR: {
      if (!draft.byShow[action.payload.id]) {
        draft.byShow[action.payload.id] = {
          ids: []
        };
      }
      draft.byShow[action.payload.id].error = action.payload.error;
      draft.byShow[action.payload.id].isLoading = false;
      break;
    }

    case FETCH_EPISODE_REQUEST: {
      if (!draft.byId[action.payload.id]) {
        draft.byId[action.payload.id] = {};
      }
      draft.byId[action.payload.id].isLoading = true;
      draft.byId[action.payload.id].error = null;
      break;
    }

    case FETCH_EPISODE_SUCCESS: {
      if (!draft.byId[action.payload.id]) {
        draft.byId[action.payload.id] = {};
      }
      draft.byId[action.payload.id].isLoading = false;
      draft.byId[action.payload.id].data = action.payload.data;
      break;
    }

    case FETCH_EPISODE_ERROR: {
      if (!draft.byId[action.payload.id]) {
        draft.byId[action.payload.id] = {};
      }
      draft.byId[action.payload.id].isLoading = false;
      draft.byId[action.payload.id].error = action.payload.error;
      break;
    }
  }
}, initialState);

export default reducer;
