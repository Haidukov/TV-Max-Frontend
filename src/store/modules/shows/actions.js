import {
  FETCH_SHOW_ERROR,
  FETCH_SHOW_REQUEST,
  FETCH_SHOW_SUCCESS
} from '../action-types';

export const fetchShowRequest = id => ({
  type: FETCH_SHOW_REQUEST,
  payload: { id }
});

export const fetchShowSuccess = (id, data) => ({
  type: FETCH_SHOW_SUCCESS,
  payload: {
    id,
    data
  }
});

export const fetchShowError = (id, error) => ({
  type: FETCH_SHOW_ERROR,
  payload: {
    id,
    error
  }
});
