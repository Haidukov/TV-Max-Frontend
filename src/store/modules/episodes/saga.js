import {
  all,
  call,
  getContext,
  put,
  takeEvery
} from 'redux-saga/effects';
import { FETCH_EPISODES_REQUEST, FETCH_EPISODE_REQUEST } from './action-types';
import {
  fetchEpisodesError,
  fetchEpisodesSuccess,
  fetchEpisodeError,
  fetchEpisodeSuccess
} from '../actions';

function* fetchShowEpisodes({ payload }) {
  try {
    const api = yield getContext('api');
    const episodes = yield call(api.getShowEpisodes, payload.id);
    yield put(fetchEpisodesSuccess(payload.id, episodes));
  } catch (e) {
    yield put(fetchEpisodesError(payload.id, e));
  }
}

function* fetchEpisode({ payload }) {
  try {
    const api = yield getContext('api');
    const episode = yield call(api.getEpisode, payload.id);
    yield put(fetchEpisodeSuccess(payload.id, episode));
  } catch (e) {
    yield put(fetchEpisodeError(payload.id, e));
  }
}

export default function* saga() {
  yield all([
    takeEvery(
      FETCH_EPISODES_REQUEST,
      fetchShowEpisodes
    ),
    takeEvery(
      FETCH_EPISODE_REQUEST,
      fetchEpisode
    )
  ]);
}
