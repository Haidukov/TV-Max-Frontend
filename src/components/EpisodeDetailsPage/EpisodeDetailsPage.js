import React, { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { useParams } from 'react-router';
import { fetchEpisodeRequest } from '../../store/modules/actions';
import { getEpisode, getEpisodeError, getEpisodeLoading } from '../../store/modules/selectors';
import DetailsPage from '../Shared/DetailsPage/DetailsPage';
import ErrorMessage from '../Shared/ErrorMessage';
import PageSpinner from '../Shared/PageSpinner';

const EpisodeDetailsPage = () => {
  const dispatch = useDispatch();
  const { episodeId } = useParams();

  const episode = useSelector(getEpisode(episodeId));
  const isLoading = useSelector(getEpisodeLoading(episodeId));
  const error = useSelector(getEpisodeError(episodeId));

  useEffect(() => {
    dispatch(fetchEpisodeRequest(episodeId));
  }, [episodeId]);

  if (isLoading) {
    return <PageSpinner />;
  }

  if (error) {
    return <ErrorMessage>Failed to load episode data</ErrorMessage>;
  }

  return (
    <DetailsPage
      summary={episode.summary}
      name={episode.name}
      imageUrl={episode.image?.medium}
    />
  );
};

export default EpisodeDetailsPage;
