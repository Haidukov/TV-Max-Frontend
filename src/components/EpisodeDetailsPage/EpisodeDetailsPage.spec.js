import React from 'react';
import { Provider } from 'react-redux';
import { MemoryRouter, Route } from 'react-router';
import { render } from '@testing-library/react';
import EpisodeDetailsPage from './EpisodeDetailsPage';
import configureStore from '../../store';
import { fetchEpisodeRequest } from '../../store/modules/actions';
import episodes from '../../fixtures/episodes.json';

const episode = episodes[0];

const renderComponent = (id, partialState = {}) => {
  const preloadedState = {
    episodes: {
      byId: {
        [id]: partialState
      },
      byShow: {}
    }
  };
  const store = configureStore(preloadedState);
  jest.spyOn(store, 'dispatch').mockImplementation(() => {});
  const wrapper = render(
    <MemoryRouter initialEntries={[`/path/${id}`]}>
      <Provider store={store}>
        <Route path='/path/:episodeId' component={EpisodeDetailsPage} />
      </Provider>
    </MemoryRouter>
  );
  return { store, wrapper };
};

const findSpinner = wrapper => wrapper.queryByRole('status');

const findErrorMessage = wrapper => wrapper.queryByRole('alert');

const findDetails = wrapper => wrapper.queryByTestId('details');

describe('<EpisodeDetailsPage />', () => {
  it('should fetch episode', () => {
    const { store } = renderComponent(episode.id);
    expect(store.dispatch).toHaveBeenCalledWith(fetchEpisodeRequest(episode.id.toString()));
  });

  describe('When loading episode', () => {
    let wrapper;
    beforeEach(() => {
      wrapper = renderComponent(episode.id, { data: null, isLoading: true }).wrapper;
    });

    it('should display loading indicator', () => {
      expect(findSpinner(wrapper)).toBeInTheDocument();
    });

    it('shouldn`t display error message', () => {
      expect(findErrorMessage(wrapper)).not.toBeInTheDocument();
    });

    it('shouldn`t display details', () => {
      expect(findDetails(wrapper)).not.toBeInTheDocument();
    });

    it('should match snapshot', () => {
      expect(wrapper.asFragment()).toMatchSnapshot();
    });
  });

  describe('When error occured', () => {
    let wrapper;
    beforeEach(() => {
      wrapper = renderComponent(episode.id, {
        data: null,
        isLoading: false,
        error: new Error('error')
      }).wrapper;
    });

    it('shouldn`t display loading indicator', () => {
      expect(findSpinner(wrapper)).not.toBeInTheDocument();
    });

    it('should display error message', () => {
      expect(findErrorMessage(wrapper)).toBeInTheDocument();
    });

    it('shouldn`t display details', () => {
      expect(findDetails(wrapper)).not.toBeInTheDocument();
    });

    it('should match snapshot', () => {
      expect(wrapper.asFragment()).toMatchSnapshot();
    });
  });

  describe('When data loaded', () => {
    let wrapper;
    beforeEach(() => {
      wrapper = renderComponent(episode.id, {
        data: episode,
        isLoading: false,
        error: null
      }).wrapper;
    });

    it('shouldn`t display loading indicator', () => {
      expect(findSpinner(wrapper)).not.toBeInTheDocument();
    });

    it('shouldn`t display error message', () => {
      expect(findErrorMessage(wrapper)).not.toBeInTheDocument();
    });

    it('should display details', () => {
      expect(findDetails(wrapper)).toBeInTheDocument();
    });

    it('should match snapshot', () => {
      expect(wrapper.asFragment()).toMatchSnapshot();
    });
  });
});
