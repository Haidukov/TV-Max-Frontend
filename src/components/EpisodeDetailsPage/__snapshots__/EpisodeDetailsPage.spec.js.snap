// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EpisodeDetailsPage /> When data loaded should match snapshot 1`] = `
<DocumentFragment>
  <main
    data-testid="details"
  >
    <h1>
      Pilot
    </h1>
    <div
      class="details"
    >
      <img
        alt="Pilot"
        class="logo"
        src="https://static.tvmaze.com/uploads/images/medium_landscape/35/89344.jpg"
      />
      <div>
        <p>
          John Reese is a former CIA operative living off the grid in NYC. When Reese is hauled into NYPD's 8th Precinct for a skirmish on the subway, he's bailed out by a man he doesn't know: Mr. Finch, a mysterious billionaire. Finch offers Reese a choice: stay on the streets and drink himself to death -- or join him in his mission to stop violent crimes before they happen. Finch has access to a secret source of information, one that identifies people who are about to be involved in violent crimes. There's just one catch: he doesn't know if these people will be the victims or the perpetrators. Skeptical but intrigued, Reese agrees to help. Together, Reese and Finch work to discover what danger is about to befall Assistant District Attorney Diane Hansen. Their actions garner the unwanted attention of NYPD homicide detective Carter -- a dedicated veteran cop who's not fond of vigilantes -- and Detective Fusco, a precinct detective who stands in the way of Reese's plans.
        </p>
      </div>
    </div>
  </main>
</DocumentFragment>
`;

exports[`<EpisodeDetailsPage /> When error occured should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="error"
    role="alert"
  >
    Failed to load episode data
  </div>
</DocumentFragment>
`;

exports[`<EpisodeDetailsPage /> When loading episode should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="spinner"
    role="status"
  >
    <span
      class="css-1k4jg2i"
    />
  </div>
</DocumentFragment>
`;
